Separate concerns relating to reflection and type information

manadart opened the pull request:
SQLair has an `internal` package, but under that a single `expr` package into which everything is lumped. 

This has allowed accretion of logic with no clear delineation of concerns, lack of encapsulation, no principle of least knowledge et al.

Here we create a new package `typeinfo` into which logic around reflection and type information is relocated. In particular:
- The exposed API surface of this logic is now apparent.
- `typeinfo.Member` (previously `typeMember`) implementations assume responsibility for returning their corresponding values within parents, and for generating scan targets. This fixes query and prepare logic so that we can rely on the indirection and eschew type switching, which violates responsibility bounds.

manadart committed:
Begins modularising the internal sub-packages.

Reflection and type information concerns are relocated to a new typeinfo
package.

Logic proviously part of prepare and query is pushed into relocated
types, which allows us to remove type-switches for typeinfo.Member
implementations.

manadart commented on internal/expr/query.go:
```diff
@@ -61,61 +63,48 @@ func (pe *PreparedExpr) Query(args ...any) (ce *QueryExpr, err error) {
-				return nil, fmt.Errorf(`map %q does not contain key %q`, outerType.Name(), tm.name)
-			}
+
+		val, err := typeMember.ValueForOuter(v)
```
This is one key part of the cleanup. If we have types with a unifying indirection, we should use that indirection to limit the knowledge required by other types.

Aflynn50 replied:
I agree, this functionality is effectively the whole raison d'etre of `typeMember`. It makes sense it should be a method of it. For the name of this function, would something like `ValueFromOuter` be better? Because we are getting the member value from the outer value.

manadart commented on internal/expr/query.go:
```diff
@@ -161,53 +151,25 @@ func (qe *QueryExpr) ScanArgs(columns []string, outputArgs []any) (scanArgs []an
-			ptrs = append(ptrs, scanVal.Addr().Interface())
-			scanProxies = append(scanProxies, scanProxy{original: outputVal, scan: scanVal, key: reflect.ValueOf(tm.name)})
+
+		if ptrs, scanProxies, err = typeMember.AddScanTarget(outputVal, ptrs, scanProxies); err != nil {
```
Here is the other.

Aflynn50 replied:
This certainly works a natural method on `typeMember`. To me it seems a bit obscure passing the `ptrs` and the `scanProxies` to the `AddScanTarget` and doing the append inside the function. It makes sense to hide the details of how the `ScanTarget` is built from the code here, but building the list seems like a local concern, not one of `typeMember`. The two arguments `ptrs` and `scanProxies` are also only used with the `append`. Could this code be:

```
if ptr, scanProxy, err = typeMember.GetScanTarget(outputVal); err != nil {
	return nil, nil, err
}
ptrs = append(ptrs, ptr)
scanProxies = append(scanProxies, scanProxy)
```
I called it `GetScanTarget` above because `ScanTarget` sounds like we are scanning a target (the name `ScanArgs` suffers from a similar problem). Any other name that is unambiguous works.

letFunny replied:
I completely agree with @Aflynn50, I think the code is clearer if we process then accumulate instead of mutating the state inside each function. Conceptually it is also not the role of a typeMember to mutate the "global" state that is used in this outer function, we are imposing state on a method that is really stateless.

SimonRichardson commented on internal/typeinfo/member.go:
```diff
@@ -0,0 +1,99 @@
+	// type member within the input outer value.
+	ValueForOuter(reflect.Value) (reflect.Value, error)
+
+	AddScanTarget(reflect.Value, []any, []ScanProxy) ([]any, []ScanProxy, error)
```
This requires documentation, along with all the public methods in this file.

Aflynn50 replied:
The documentation for this function might be a better place to document the behavior on NULL than where it is currently (in the `AddScanTarget` function of `structField`). The `AddScanTarget` method of `structField` would also work.

SimonRichardson commented on internal/typeinfo/scan.go:
```diff
@@ -0,0 +1,25 @@
+
+// ScanProxy is a shim for scanning query results
+// into types for which we have information.
+type ScanProxy struct {
```
This feels like it's doing too much. I'm not suggesting fixing it now, as you're doing a lot of mechanical changes, but we should have a `MapSetterProxy` and a `ValueSetterProxy`. Only `mapKey` returns the key, so that should return a `MapSetterProxy`, then `structField` can just return a `ValueSetterProxy`.  The `AddScanTarget` should take and return a `Proxy` interface.

```go
type Proxy interface {
    SetValue()
}
```

That way we remove the need for the if statement in `OnSuccess`, in addition to changing the name which doesn't tell you the intent of what it's actually doing.

manadart replied:
I'm trying to limit this to a lift and shift. 

I actually started to do some other rework, and the change-set got unwieldy very fast. Once we have things a little better organised, I imagine we'll reconsider these implementations wholesale in the context of the features we need to add.

letFunny replied:
I think whether to dispatch or pattern match (see my general comment) comes down to personal opinion in the end, but I also think no paradigm should be absolute. As I said, my fear is that we do not take into account the original design and its quest for "simplicity".

In this concrete case we would be creating one file, one interface, three struct types and three methods to refactor a <20 line function. Is there a reason for that, are we envisioning that we will support more data types or that we need this added complexity?

letFunny replied:
> That way we remove the need for the if statement in OnSuccess, in addition to changing the name which doesn't tell you the intent of what it's actually doing.

We would not be removing the if though, we would be using method dispatch do it instead, effectively we would only be moving the conditional elsewhere.

Aflynn50 commented on internal/expr/query.go:
```diff
@@ -161,53 +151,25 @@ func (qe *QueryExpr) ScanArgs(columns []string, outputArgs []any) (scanArgs []an
-				}
-				sp.original.Set(val)
-			}
+			sp.OnSuccess()
```
Lovely

Aflynn50 commented on internal/typeinfo/scan.go:
```diff
@@ -0,0 +1,25 @@
+	key      reflect.Value
+}
+
+func (sp ScanProxy) OnSuccess() {
```
Documentation needed to make it clear this needs to be run after `rows.Scan`.

Aflynn50 commented on internal/typeinfo/scan.go:
```diff
@@ -0,0 +1,25 @@
+import "reflect"
+
+// ScanProxy is a shim for scanning query results
+// into types for which we have information.
```
The last part of the sentence is not very clear, maybe add that `ScanProxy` is used when scanning directly into the target is not possible/desired

Aflynn50 commented on internal/typeinfo/member.go:
```diff
@@ -0,0 +1,99 @@
+	return val, nil
+}
+
+func (mk *mapKey) AddScanTarget(outVal reflect.Value, ptrs []any, proxies []ScanProxy) ([]any, []ScanProxy, error) {
```
It would be good to note the reason that we need the `ScanProxy` here, namely that Go does not expose pointers to map values (this would maybe be an internal comment rather than on the public function).

Aflynn50 reviewed (commented):
This feels like a very natural change, it makes the code a lot cleaner. It also great to see some of the small things tidied too like variable declarations up that have been bugging me ever since I started trying to be consistent.

letFunny reviewed (commented):
In general I quite like the changes to the code, especially:
* The idea of creating another internal package for `typeinfo` if that is going to lead to better encapsulation especially by restricting the use of reflect only to that module.
* The creation of `typeMember.ValueForOuter(...)` which seems to be the correct abstraction here.

Apart from that, I feel we should have a conversation about the goals of the project and the paradigm that we want to follow. We briefly touched on that in an offline discussion and it comes down to method dispatching and encapsulation vs pattern matching. The project has been mostly following the latter by using interfaces as enums and by performing pattern matching using switch statements on those interfaces. We seem to be now moving towards method dispatching and encapsulation which is all right, it comes down to personal preference after all. My only fear is that we do not take into account the original design of SQLair and its shared vision.

Specifically, one of the main objectives of SQLair was to be simple and straightforward to read and I am not sure that continues to be a shared goal (see [this comment](https://github.com/canonical/sqlair/pull/121/files#r1395838194) or [this one](https://github.com/canonical/sqlair/pull/121/files#r1395715007)). For this reason, if we decide on changing the paradigm to encapsulation and indirection, I think it is important that we continue to balance it against our explicit goal of simplicity and explicitness.

Lastly, if we decide to sacrifice this simplicity for better modularization or any other goal, I will commit fully to the new values, but I think it is a conversation that we need to have.

manadart committed:
Changes AddScanTarget to GetScanTarget and modifies its behaviour so as
not to append to input slices. Those slices are accumulated by the
caller, which makes for greater simplicity.

Better documentation and an additional method name change accompany.

Aflynn50 reviewed (approved):
LGTM
